web-api/src/mind.rs

```diff
@@ -231,6 +299,122 @@ where
+        .writer(writer)
+        .begin_nd()?;
+    writer.extend(c_order_items)?;
+    writer.finish()
+}
+#[doc = "
+
+# Panics
+
+The function panics if the provided filenames are not correct.
+"]
+pub async fn run_persona_based_benchmark() -> Result<(), Error> {
+    let document_provider = DocumentProvider::new("news.tsv")?;
+    let state = State::new(Storage::default()).unwrap();
+    // load documents from document provider to state
+    state
+        .insert(document_provider.documents.values().cloned().collect_vec())
+        .await
+        .unwrap();
+    let benchmark_config = PersonaBasedConfig::default();
+    // create 3d array of zeros with shape (nranks, iterations, n_documents)
+    let mut results = Array3::<f32>::zeros([
+        benchmark_config.nranks.len(),
+        benchmark_config.iterations,
+        benchmark_config.n_documents,
+    ]);
+
+    let mut rng = thread_rng();
+
+    let users_interests = Users::new("user_categories.json")?;
+    for (idx, (user_id, interests)) in users_interests.iter().enumerate() {
+        let interesting_documents = document_provider.get_all_interest(interests);
+        let ids_of_documents_to_prepare = interesting_documents
+            .choose_multiple(&mut rng, benchmark_config.amount_of_doc_used_to_prepare)
+            .map(|doc| doc.id.clone())
+            .collect_vec();
+        // prepare reranker by interacting with documents to prepare
+        state
+            .interact(user_id, &ids_of_documents_to_prepare)
+            .await
+            .unwrap();
+        // perform iterations
+        for iter in 0..benchmark_config.iterations {
+            // get personalised documents from state
+            let personalised_documents = state
+                .personalize(
+                    user_id,
+                    PersonalizeBy::KnnSearch(benchmark_config.n_documents),
+                )
+                .await
+                .unwrap();
+            // get documents based on the personalised documents ids
+            let documents = personalised_documents
+                .iter()
+                .map(|id| document_provider.get(id).unwrap())
+                .collect_vec();
+            // score documents based on the user's interests
+            let scores = score_documents(&documents, interests);
+            //save scores to results array
+            for (i, score) in scores.iter().enumerate() {
+                results[[idx, iter, i]] = *score;
+            }
+
+            // interact with some of the retrieved documents based on whether the document is interesting to the user and probability of clicking
+            // interact with documents
+            state
+                .interact(
+                    user_id,
+                    &personalised_documents
+                        .iter()
+                        .zip(scores.iter())
+                        .filter(|(_, &score)| {
+                            (score - 2.0).abs() < 0.001
```

Comment:
Could we not just check if the score is 2?

Reply:
exact float comparrison is usually error prone because of float rounding precision errors